### PR TITLE
Azure Cognitive Search connector for Semantic Memory

### PR DESCRIPTION
**Pull request by dluc**
### Motivation and Context

Azure Cognitive Search is a managed service on Azure that allows to store data and search it using natural language.

### Description

Add a new Semantic Memory Connector leveraging Azure Cognitive Search. Update the syntax example to show how it seamlessly provides `ISemanticTextMemory` API without the need to use embeddings.

### CODE REVIEW

File: dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemory.cs
Comment (stephentoub) [inline dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemory.cs:205]: In what situation will `client == null`?

File: dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemory.cs
Comment (stephentoub) [inline dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemory.cs:211]: Two concurrent calls to GetSearchClientAsync could end up getting a different SearchClient instance.  Is that ok?  With small tweaks here you could ensure the same instance is returned regardless of such race conditions.

File: dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemory.cs
Comment (stephentoub) [inline dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemory.cs:296]: The Regex static methods are easy to use but have hidden costs.  It'd be better to move it to a field and then use the instance Replace method, ala https://github.com/microsoft/semantic-kernel/pull/728.

File: dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemoryException.cs
Comment (stephentoub) [inline dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemoryException.cs:32]: This should either be public or deleted.

File: dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemory.cs
Comment (dluc) [reply dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemory.cs:211]: it would be fine, but I'll double check for improvements

File: dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemoryException.cs
Comment (dluc) [reply dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemoryException.cs:32]: IIRC the build throws warnings if it's not defined. We make it private to stop people from raising exceptions without providing useful information.

Comment (stephentoub) [reply dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemoryException.cs:32]: That's because there's an analyzer enabled that's checking that condition; its actual intent is to ensure that such ctors exist and are public. If weet don't want to expose such ctors, the analyzer should be disabled, or if we don't want to disable it everywhere, it should instead be suppressed here. Adding such a private ctor is an expebsive way to suppress the analyzer :-)

File: dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemory.cs
Comment (adrianwyatt) [inline dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemory.cs:57]: Do we want to do any parameter validation?

File: dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemory.cs
Comment (adrianwyatt) [inline dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemory.cs:259]: Should we check `result.Value.Results.Count > 0` as well?

File: dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemory.cs
Comment (adrianwyatt) [inline dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemory.cs:296]: Could you add a comment describing the intent of the regex? For example: replace any whitespace, backslash, forward slash, period, underscore or colon with a dash.

File: dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchRecord.cs
Comment (adrianwyatt) [inline dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchRecord.cs:8]: XMLDoc throughout would be lovely.

**Review by adrianwyatt — CHANGES_REQUESTED** [review]

File: dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemory.cs
Comment (dluc) [reply dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemory.cs]: yeah that's after the other PR merge. I had these fixed and then reverted and now I'm fixing again lol

File: dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemory.cs
Comment (dluc) [reply dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemory.cs:267]: it's very unlikely, but I added the check

File: dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchRecord.cs
Comment (dluc) [reply dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchRecord.cs:8]: I added to all the public methods and to the non-obvious ones. Let me know if I missed some :-)

File: dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemory.cs
Comment (dluc) [reply dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemory.cs:226]: I can integrate the optimization later, I wouldn't block the PR on this. Could you share the code?

**Review by hathind-ms — APPROVED** [review]

File: dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemory.cs
Comment (adrianwyatt) [reply dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemory.cs:60]: I see - we're just passing it down the line.

**Review by adrianwyatt — APPROVED** [review]

File: dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemoryException.cs
Comment (dluc) [reply dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemoryException.cs]: solved

**Pull request merged**
